test/python/dialects/indexing/dialect.py

```diff
@@ -474,3 +477,153 @@ def testCanonicalizeTupleIndexCastListLiteral():
+
+
+# CHECK-LABEL: TEST: testAdvancedIndexing
+@run
+def testAdvancedIndexing():
+  index = IndexType.get()
+  f32 = F32Type.get()
+  with mlir_mod_ctx() as module:
+    ten = Tensor.empty((10, 10, 10, 10), f32)
+    # CHECK: Tensor(%[[TEN:.*]], tensor<10x10x10x10xf32>)
+    print(ten)
+
+    w = ten[[[0], [1]], :, :, :]
+    idx_tensor_operand = Tensor(w.owner.operands[1])
+    # CHECK: %[[IDXTEN:.*]] tensor<2x1xindex> True
+    print(idx_tensor_operand.get_name(), idx_tensor_operand.type,
+          idx_tensor_operand.is_constant())
+    # CHECK{LITERAL}: [[0], [1]]
+    print(get_array_on_one_line(idx_tensor_operand.literal_value))
+    # CHECK: %{{.*}} = indexing.gather %[[TEN]][%[[IDXTEN]]] gather_dims([0]) unique : (tensor<10x10x10x10xf32>, tensor<2x1xindex>) -> tensor<2x10x10x10xf32>
+    print(w.owner)
+
+    w = ten[[[0], [1]], [[0], [1]], :, :]
+    idx_tensor_operand = Tensor(w.owner.operands[1])
+    # CHECK: %[[IDXTEN:.*]] tensor<2x2xindex> True
+    print(idx_tensor_operand.get_name(), idx_tensor_operand.type,
+          idx_tensor_operand.is_constant())
+    # CHECK{LITERAL}: [[0 0], [1 1]]
+    print(get_array_on_one_line(idx_tensor_operand.literal_value))
+    # CHECK: %{{.*}} = indexing.gather %[[TEN]][%[[IDXTEN]]] gather_dims([0, 1]) unique : (tensor<10x10x10x10xf32>, tensor<2x2xindex>) -> tensor<2x10x10xf32>
+    print(w.owner)
+
+    w = ten[[[0], [1]], :, [[0], [1]], :]
+    idx_tensor_operand = Tensor(w.owner.operands[1])
+    # CHECK: %[[IDXTEN:.*]] tensor<2x2xindex> True
+    print(idx_tensor_operand.get_name(), idx_tensor_operand.type,
+          idx_tensor_operand.is_constant())
+    # CHECK{LITERAL}: [[0 0], [1 1]]
+    print(get_array_on_one_line(idx_tensor_operand.literal_value))
+    # CHECK: %{{.*}} = indexing.gather %[[TEN]][%[[IDXTEN]]] gather_dims([0, 2]) unique : (tensor<10x10x10x10xf32>, tensor<2x2xindex>) -> tensor<2x10x10xf32>
+    print(w.owner)
+
+    # ensure that the indices aren't unique because the checks below all except non-unique
+    idx_tensor = np.random.randint(low=0, high=10, size=(5, 6, 1))
+    while len(np.unique(idx_tensor)) == idx_tensor.size:
+      idx_tensor = np.random.randint(low=0, high=10, size=(5, 6, 1))
+    idx_tensor = Tensor(idx_tensor, dtype=index)
+    # CHECK: %[[IDXTEN:.*]] tensor<5x6x1xindex> True
+    print(idx_tensor.get_name(), idx_tensor.type, idx_tensor.is_constant())
+    # CHECK: %[[IDXTEN:.*]] = arith.constant dense<{{.*}}> : tensor<5x6x1xindex>
+    print(idx_tensor.owner)
+    # CHECK: True
+    print(np.array(idx_tensor.literal_value).dtype == np.int64)
+
+    w = ten[idx_tensor, ...]
+    idx_tensor_operand = Tensor(w.owner.operands[1])
+    # CHECK: %[[IDXTEN:.*]] tensor<5x6x1xindex> True
+    print(idx_tensor_operand.get_name(), idx_tensor_operand.type,
+          idx_tensor_operand.is_constant())
+    # CHECK: %{{.*}} = indexing.gather %[[TEN]][%[[IDXTEN]]] gather_dims([0]) : (tensor<10x10x10x10xf32>, tensor<5x6x1xindex>) -> tensor<5x6x10x10x10xf32>
```

Comment:
now that I look at this test I am wondering whether you really want 10x10x10x10 
can you use e.g. `7x77x777x7777` instead so we disambiguate between dimensions in the type?

Reply:
very good idea